src/cluster.c

```diff
@@ -6102,6 +6102,26 @@ int clusterRedirectBlockedClientIfNeeded(client *c) {
     return 0;
 }
 
+/* server.cluster->slots_to_keys
+ *     Represents the slots to keys map of current database being served.
+ *     We add/remove/read from it when there's no replication in progress or when
+ *     the replication is not async.
+ * 
+ * server.cluster->slots_to_keys_tempdb
+ *     Is only relevant when there's an async diskless replication in progress.
+ *     During async loading that is where we write before it eventually becomes the
+ *     main slots_to_keys.
+ *
+ * The purpose of this method is to facilitate correct selection of slots to keys object.
+ * When we simply need to read it with the purpose of serving data and not changing it,
+ * server.cluster->slots_to_keys must be accessed instead of calling this method. */
+clusterSlotsToKeysData *slotToKeyGetForChanging(void) {
```

Comment:
I think that the change in your last commit is technically correct, but I must say I don't like it too much. 
I think it would be better to move this structure to the redisDb struct (to be allocated only when cluster is enabled). 
This way dbAdd would handle it correctly, and the tempDB we allocate during replication would be enough (no need for an extra member in the cluster struct). 
I think it's OK for server.h to be aware of the slots count, it doesn't have to be abstracted in cluster.h. 
@madolson WDYT?

@eduardobr btw, maybe we better block async loading of someone enabled a writable replica? 
I don't usually care for writable replica, enabling it is an invitation but bugs and inconsistencies.
But looking at the current code ("forChanging"), it'll obviously do the wrong thing..

Reply:
I started with that solution, but then felt forced to have it in a public place in order to not flood methods with tempDb argument. But also agree it would be more self-contained.

About writable replicas, they are more for caching calculated results, like those commands with STORE flag and things like that, right? I'm indifferent, your call ;)
BTW, now that we just need minor changes, please feel free to make commits. Unfortunately I'll be short on time during next week, then I'm more available from the 26th.

As I see, we have:
- This possible refactoring for the placing temporary slots_to_keys in different location
- Adjust some async_loading conditionals as mentioned here https://github.com/redis/redis/pull/9323#discussion_r709567812
- Maybe better support deprecated module api https://github.com/redis/redis/pull/9323#discussion_r708332098, which maybe is just about doing this: https://github.com/redis/redis/pull/9323#discussion_r709087671

Reply:
I like the idea of putting the slot-to-key mapping (pointer) inside the redisDb struct. Then, a temp db can just be a redisDb and we don't need a specific tempDb struct.

The slot-to-key mapping can still be opaque and delegated from db to cluster if we like, independently of that.